File: src/pyftp_server/wormhole/crypto.py
"""
crypto.py
=========
虫洞端到端加密(AES-256-GCM)，被 sync.py(FTP 中转) 和 p2p.py(局域网直连) 共用。

加密文件格式: magic"WHE1" + salt(16B) + nonce(12B) + AES-GCM密文(含校验标签)
密钥由口令经 PBKDF2-HMAC-SHA256 派生，salt/nonce 每个文件随机。
传输全程只见密文——即使中转服务器/网络窃听者都看不到内容。
"""

from __future__ import annotations
import os

_MAGIC = b"WHE1"


def _derive_key(secret: str, salt: bytes) -> bytes:
    import hashlib
    return hashlib.pbkdf2_hmac("sha256", secret.encode("utf-8"), salt, 100_000, dklen=32)


def encrypt(secret: str, plain: bytes) -> bytes:
    """加密明文 bytes，返回 magic+salt+nonce+密文。"""
    from cryptography.hazmat.primitives.ciphers.aead import AESGCM
    salt, nonce = os.urandom(16), os.urandom(12)
    return _MAGIC + salt + nonce + AESGCM(_derive_key(secret, salt)).encrypt(nonce, plain, None)


def decrypt(secret: str, blob: bytes) -> bytes | None:
    """解密；不是加密格式或口令不对返回 None。"""
    from cryptography.hazmat.primitives.ciphers.aead import AESGCM
    from cryptography.exceptions import InvalidTag
    if not is_encrypted(blob) or len(blob) < 4 + 16 + 12 + 16:
        return None
    salt, nonce, ct = blob[4:20], blob[20:32], blob[32:]
    try:
        return AESGCM(_derive_key(secret, salt)).decrypt(nonce, ct, None)
    except InvalidTag:
        return None


def is_encrypted(blob: bytes) -> bool:
    """检查数据是否以虫洞加密 magic 开头。"""
    return blob[:4] == _MAGIC

File: android/app/src/main/java/com/rexvane/wormhole/p2p/Crypto.kt
package com.rexvane.wormhole.p2p

import javax.crypto.Cipher
import javax.crypto.SecretKeyFactory
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.PBEKeySpec
import javax.crypto.spec.SecretKeySpec
import java.security.SecureRandom

/**
 * AES-256-GCM 端到端加密, 与桌面版 Python crypto.py 完全兼容。
 *
 * 格式: [4B magic "WHE1"] [16B salt] [12B nonce] [ciphertext + 16B GCM tag]
 * 密钥派生: PBKDF2-HMAC-SHA256, 100000 次迭代, 256-bit。
 */
object Crypto {
    private val MAGIC = "WHE1".toByteArray(Charsets.US_ASCII)
    private const val SALT_LEN = 16
    private const val NONCE_LEN = 12
    private const val ITERATIONS = 100_000
    private const val KEY_LEN = 256     // bits
    private const val TAG_LEN = 128     // bits (GCM auth tag)

    fun isEncrypted(blob: ByteArray): Boolean {
        return blob.size >= 4 + SALT_LEN + NONCE_LEN && blob.copyOfRange(0, 4).contentEquals(MAGIC)
    }

    fun encrypt(secret: String, plain: ByteArray): ByteArray {
        val sr = SecureRandom()
        val salt = ByteArray(SALT_LEN).also { sr.nextBytes(it) }
        val nonce = ByteArray(NONCE_LEN).also { sr.nextBytes(it) }
        val key = deriveKey(secret, salt)
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.ENCRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(TAG_LEN, nonce))
        val ct = cipher.doFinal(plain)
        // 拼接: magic + salt + nonce + ciphertext(含 tag)
        return MAGIC + salt + nonce + ct
    }

    fun decrypt(secret: String, blob: ByteArray): ByteArray? {
        if (!isEncrypted(blob)) return null
        val salt = blob.copyOfRange(4, 4 + SALT_LEN)
        val nonce = blob.copyOfRange(4 + SALT_LEN, 4 + SALT_LEN + NONCE_LEN)
        val ct = blob.copyOfRange(4 + SALT_LEN + NONCE_LEN, blob.size)
        val key = deriveKey(secret, salt)
        return try {
            val cipher = Cipher.getInstance("AES/GCM/NoPadding")
            cipher.init(Cipher.DECRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(TAG_LEN, nonce))
            cipher.doFinal(ct)
        } catch (e: Exception) {
            null
        }
    }

    private fun deriveKey(secret: String, salt: ByteArray): ByteArray {
        val spec = PBEKeySpec(secret.toCharArray(), salt, ITERATIONS, KEY_LEN)
        val factory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256")
        return factory.generateSecret(spec).encoded
    }
}
